libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp

```diff
@@ -70,3 +75,132 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
+        smart_batt_info.got_cycle_count ? smart_batt_info.cycles : UINT16_MAX,                // Charge/discharge cycle count. UINT16_MAX: field not provided.
+        smart_batt_info.serial_number,      // Serial number in ASCII characters, 0 terminated. All 0: field not provided.
+        smart_batt_info.product_name,       // Static device name. Encode as manufacturer and product names separated using an underscore.
+        0,                                  // (not provided) [g] Battery weight. 0: field not provided.
+        UINT16_MAX,                         // (not provided) [mV] Minimum per-cell voltage when discharging. If not supplied set to UINT16_MAX value.
+        UINT16_MAX,                         // (not provided) [mV] Minimum per-cell voltage when charging. If not supplied set to UINT16_MAX value.
+        UINT16_MAX,                         // (not provided) [mV] Minimum per-cell voltage when resting. If not supplied set to UINT16_MAX value.
+        smart_batt_info.got_design_voltage ? (smart_batt_info.design_voltage / smart_batt_info.cells_in_series * 1000.0) : 0, // [mV] Maximum per-cell voltage when charged. 0: field not provided.
+        smart_batt_info.cells_in_series,    // Number of battery cells in series. 0: field not provided.
+        0,                                  // (not provided) [mA] Maximum pack discharge current. 0: field not provided.
+        0,                                  // (not provided) [mA] Maximum pack discharge burst current. 0: field not provided.
+        smart_batt_info.manufacture_date);  // Manufacture date (DD/MM/YYYY) in ASCII characters, 0 terminated. All 0: field not provided.
+
+    return true;
+#endif // !defined(HAL_NO_GCS)
+#endif // HAL_SMART_BATTERY_INFO_ENABLED
+
+    return false;
+}
+
+    // Smart battery info. helper to gather data for logging and mavlink
+    // returns false if cells_in_series is not available
+bool AP_BattMonitor_Backend::get_smart_batt_info(Smart_Batt_Info &smart_batt_info) const
+{
+#if HAL_SMART_BATTERY_INFO_ENABLED
+
+    smart_batt_info.got_cells_in_series = get_cells_in_series(smart_batt_info.cells_in_series);
+
+    // get_cells_in_series only returns false if cells_in_series is not available
+    if (!smart_batt_info.got_cells_in_series) {
+       return false;
+    }
+
+    smart_batt_info.got_cycle_count = get_cycle_count(smart_batt_info.cycles);
+    smart_batt_info.got_capacity_design = get_design_capacity(smart_batt_info.capacity_design);
+    smart_batt_info.got_capacity_full = get_full_charge_capacity(smart_batt_info.capacity_full);
+    smart_batt_info.got_design_voltage = get_design_voltage(smart_batt_info.design_voltage);
+    get_product_name(smart_batt_info.product_name, ARRAY_SIZE(smart_batt_info.product_name));
+    get_manufacture_date(smart_batt_info.manufacture_date,  ARRAY_SIZE(smart_batt_info.manufacture_date));
+
+    const int32_t ser_num =  _params._serial_number;
```

Comment:
Don't we want to cast this to `uint32_t`?  It is an artifact of our param system that we can't distinguish signed/unsigned well

Reply:
`_params._serial_number` is -1 until the smartbatt gets the serial number. And I need to check if that is the case to null terminate the string if no serial number was read.  Is this type of check ok `_params._serial_number != -1`?